src/org/traccar/database/DataManager.java

```diff
@@ -115,6 +124,73 @@ private void initDatabase() throws Exception {
         }
     }
 
+    public static String constructObjectQuery(String action, Class<?> clazz, boolean additional) {
+        switch (action) {
+            case ACTION_INSERT:
+            case ACTION_UPDATE:
+                StringBuilder result = new StringBuilder();
+                StringBuilder fields = new StringBuilder();
+                StringBuilder values = new StringBuilder();
+
+                Set<Method> methods = new HashSet<>(Arrays.asList(clazz.getMethods()));
+                methods.removeAll(Arrays.asList(Object.class.getMethods()));
+                methods.removeAll(Arrays.asList(BaseModel.class.getMethods()));
+                for (Method method : methods) {
+                    if (method.getName().startsWith("get") && method.getParameterTypes().length == 0
+                            && (additional ? method.isAnnotationPresent(QueryAdditional.class)
+                                : !method.isAnnotationPresent(QueryIgnore.class)
+                                && !method.isAnnotationPresent(QueryAdditional.class))) {
```

Comment:
User has two stages of update, `hashedPassword`/`salt` is additional other fields are standard.
Device also have two stages, `lastUpdate` is additional, other fields are standard.

I did not check `@QueryIgnore` if additional is true, because I hope we will not use them in combination, it is meaningless, moreover it will cause problems because `QueryBuilder` will not set such fields in statement.

It was only way that I could invent. If you have another idea, please share.